File: No-lineal structure/Structure in js/Arboles/Merkle/merkle.js

class Node {
    constructor(left, right, value, content) {
        this.left = left;
        this.right = right;
        this.value = value;
        this.content = content;
    }

    static hash(val) {
        return require("crypto").createHash("sha256").update(val).digest("hex");
    }

    static doubleHash(val) {
        return this.hash(this.hash(val));
    }
}

export class MerkleTree {
    constructor(values) {
        this.__values = this.calculateLength(values);
        this.__buildTree(this.__values);
    }
   
    __buildTree(values) {
        let leaves = values.map(e => new Node(null, null, Node.doubleHash(e), e));
        if (leaves.length % 2 == 1) {
            leaves.push(leaves[leaves.length - 1]); // duplicate last elem if odd number of elements
        }
        this.root = this.__buildTreeRec(leaves);
    }

    __buildTreeRec(nodes) {
        let half = Math.floor(nodes.length / 2);
        if(nodes.length == 2) {
            return new Node(nodes[0], nodes[1], Node.doubleHash(nodes[0].value + nodes[1].value), nodes[0].content + nodes[1].content);
        }
        let left = this.__buildTreeRec(nodes.slice(0, half));
        let right = this.__buildTreeRec(nodes.slice(half));
        let value = Node.doubleHash(left.value + right.value);
        let content = left.content + right.content;
        return new Node(left, right, value, content);
    }
    calculateLength(values) {
        let res = [];
        for(let i = 0; i < values.length; i++) {
            let value = values[i];
            let string = value.toString();
            let length = string.length;
            let result = string + length.toString();
            res.push(result);
        }
        return res;
    }
    printTree() {
        this.__printTreeRec(this.root);
    }
    __printTreeRec(node) {
        if(node != null) {
            console.log(node.content);
            this.__printTreeRec(node.left);
            this.__printTreeRec(node.right);
        }
    }

    graficar(nombre) {
        let cadena = "";
        cadena += "digraph G { \n";
        cadena += "rankdir=TB; \n";
        cadena += "label=\"";
        cadena += nombre + "\"\n";
        cadena += "fontname=\"Arial Black\"\n" + "fontsize=\"25pt\"\n";
        cadena += "node[color=\"blue\",style=\"rounded,filled\",fillcolor=lightgray, shape=record, fontname=\"Arial\"];\n";
        cadena +=  this.graficadora(this.root);
        cadena += "} \n";
        let dotX = "MerkleTree_{}.dot".format("Base");
        let file = open(dotX, "w");
        file.write(cadena);
        file.close();
        let result = "MerkleTree_{}.png".format(nombre);
        system("dot -Tpng " + dotX + " -o " + result);
        let commando = "xdg-open " + result;
        system(commando);
        //print(cadena)
    }

    isNumeric(s) {
        try {
            complex(s);
            return true;
        } catch (e) {
            return false;
        }
    }

    graficadora(node) {
        let cadena = "";
        
        if((node.right === null) && (node.left === null)) {
            cadena = "\"nodo" + node.value + "\"" + "[label =\"" + node.value + "\"]; \n";
    
            if (this.isNumeric(node.content)) {
                cadena += "\"nodo" + node.content + "\"" + "[label =\"" + str(-1) + "\"]; \n";
                cadena += "\"nodo" + node.value + "\"" + "->" + "\"nodo" + node.content + "\"[dir=back];\n";
            } else {
                cadena += "\"nodo" + node.content + "\"" + "[label =\"" + node.content + "\"]; \n";
                cadena += "\"nodo" + node.value + "\"" + "->" + "\"nodo" + node.content + "\"[dir=back];\n";
            }
    
        } else {
            cadena = "\"nodo" + node.value + "\"" + "[label =\"<C0>|" + node.value + "|<C1> \"]; \n";
        }
        
        if(node.left !== null) {
            cadena += this.graficadora(node.left) +  "\"nodo" + node.value + "\":C0->\"nodo" + node.left.value + "\"[dir=back]\n";
        }
        if(node.right !== null) {
            cadena += this.graficadora(node.right) + "\"nodo" + node.value +  "\":C1->\"nodo" + node.right.value + "\"[dir=back]\n";
        }
        
        return cadena;
    }

    getRootHash = () => {
        return this.root.value
    }

    calculateLength(lts) {
        var k = 0;
        var bin = 2;
        while(lts.length >= 2 ** k){
            k += 1; //sigue creciendo
        }
        bin = bin ** k;
        var nuevoLarge = bin-lts.length;
        for(var x = 0; x < nuevoLarge; x++){
            lts.push((-1) * (x+1));
        }
        console.log(lts);
        return lts;
    }
}

File: No-lineal structure/Structure in js/MerkleBlocchain/Merkle.py
from typing import List
import hashlib
from os import system
class Node:
    def __init__(self, left, right, value: str, content)-> None:
        self.left: Node = left
        self.right: Node = right
        self.value = value
        self.content = content

    @staticmethod
    def hash(val: str)-> str:
        return hashlib.sha256(val.encode('utf-8')).hexdigest()

    @staticmethod
    def doubleHash(val: str)-> str:
        return Node.hash(Node.hash(val))

class MerkleTree:
    def __init__(self, values: List[str])-> None:
        self.__values = self.calculateLength(values)
        self.__buildTree(self.__values)

    def __buildTree(self, values: List[str])-> None:
        leaves: List[Node] = [Node(None, None, Node.doubleHash(e), e) for e in values]
        if len(leaves) % 2 == 1:

            leaves.append(leaves[-1:][0]) # duplicate last elem if odd number of elements
        self.root: Node = self.__buildTreeRec(leaves)

    def __buildTreeRec(self, nodes: List[Node])-> Node:
        half: int = len(nodes) // 2

        if len(nodes) == 2:
            return Node(nodes[0], nodes[1], Node.doubleHash(nodes[0].value + nodes[1].value), nodes[0].content + nodes[1].content)

        left: Node = self.__buildTreeRec(nodes[:half])
        right: Node = self.__buildTreeRec(nodes[half:])
        value: str = Node.doubleHash(left.value + right.value)
        content: str = left.content + right.content
        return Node(left, right, value, content)

    def printTree(self)-> None:
        self.__printTreeRec(self.root)

    def __printTreeRec(self, node:Node)-> None:
        if node != None:
            print(node.content)
            self.__printTreeRec(node.left)
            self.__printTreeRec(node.right)
    
    def graficar(self, nombre):
        #print(self.__values)
        cadena = ""
        cadena += "digraph G { \n"
        cadena += "rankdir=TB; \n"
        cadena += "label=\""
        cadena += nombre + "\"\n"
        cadena += "fontname=\"Arial Black\"\n" + "fontsize=\"25pt\"\n"
        cadena += "node[color=\"blue\",style=\"rounded,filled\",fillcolor=lightgray, shape=record, fontname=\"Arial\"];\n"
        cadena +=  self.graficadora(self.root)
        cadena += "} \n"
        dotX = "MerkleTree_{}.dot".format("Base")
        file = open(dotX, "w")
        file.write(cadena)
        file.close()
        result = "MerkleTree_{}.png".format(nombre)
        system("dot -Tpng " + dotX + " -o " + result)
        commando = "xdg-open " + result
        system(commando)
        #print(cadena)
    
    def isNumeric(self, s):
        try:
            complex(s)
            return True
        except ValueError:
            return False

    def graficadora(self, node:Node):
        cadena = ""
        
        if((node.right is None) and (node.left is None)):
            cadena = "\"nodo" + node.value + "\"" + "[label =\"" + node.value + "\"]; \n"

            if (self.isNumeric(node.content)):
                cadena += "\"nodo" + node.content + "\"" + "[label =\"" + str(-1) + "\"]; \n"
                cadena += "\"nodo" + node.value + "\"" + "->" + "\"nodo" + node.content + "\"[dir=back];\n"
            else:
                cadena += "\"nodo" + node.content + "\"" + "[label =\"" + node.content + "\"]; \n"
                cadena += "\"nodo" + node.value + "\"" + "->" + "\"nodo" + node.content + "\"[dir=back];\n"

        else:
            cadena = "\"nodo" + node.value + "\"" + "[label =\"<C0>|" + node.value + "|<C1> \"]; \n"
        
        if(node.left is not None):
            cadena += self.graficadora(node.left) +  "\"nodo" + node.value + "\":C0->\"nodo" + node.left.value + "\"[dir=back]\n"
        if(node.right is not None):
            cadena += self.graficadora(node.right) + "\"nodo" + node.value +  "\":C1->\"nodo" + node.right.value + "\"[dir=back]\n"
        
        return cadena


    def getRootHash(self)-> str:
        return self.root.value
    
    def calculateLength(self, lts:list)-> list:
        k = 0
        bin = 2
        while(len(lts) >= 2 ** k):
            k += 1 #sigue creciendo
        bin = bin ** k
        nuevoLarge = bin-len(lts)
        for x in range(0, nuevoLarge):
            lts.append(str((-1) * (x+1)))
        print(lts)
        return lts